// react-snippets/Composer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Await = require('./Await');

var _Await2 = _interopRequireDefault(_Await);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var func = _react.PropTypes.func,
    arrayOf = _react.PropTypes.arrayOf;

var Composer = function (_Component) {
    (0, _inherits3.default)(Composer, _Component);

    function Composer() {
        var _ref;

        var _temp, _this, _ret;

        (0, _classCallCheck3.default)(this, Composer);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Composer.__proto__ || (0, _getPrototypeOf2.default)(Composer)).call.apply(_ref, [this].concat(args))), _this), _this.onStart = function (resolve, reject) {
            var _this$props = _this.props,
                actions = _this$props.actions,
                onStart = _this$props.onStart;

            if (onStart) {
                onStart(resolve, reject);
            }
            _promise2.default.all(actions.map(function (action) {
                return action();
            })).then(resolve).catch(reject);
        }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
    }

    (0, _createClass3.default)(Composer, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                renderComplete = _props.renderComplete,
                renderPending = _props.renderPending,
                onSuccess = _props.onSuccess,
                onError = _props.onError,
                onCancel = _props.onCancel;

            return _react2.default.createElement(_Await2.default, {
                renderComplete: renderComplete,
                renderPending: renderPending,
                onStart: this.onStart,
                onSuccess: onSuccess,
                onError: onError,
                onCancel: onCancel
            });
        }
    }]);
    return Composer;
}(_react.Component);

Composer.propTypes = {
    actions: arrayOf(func).isRequired,

    renderComplete: func.isRequired,
    renderPending: func.isRequired,

    onStart: func,
    onSuccess: func,
    onError: func,
    onCancel: func
};
exports.default = Composer;
