// react-snippets/EventListener.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Empty = require('./Empty');

var _Empty2 = _interopRequireDefault(_Empty);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

if (!Element.prototype.matches) {
    Element.prototype.matches = Element.prototype.matchesSelector || Element.prototype.mozMatchesSelector || Element.prototype.msMatchesSelector || Element.prototype.oMatchesSelector || Element.prototype.webkitMatchesSelector || function (s) {
        var matches = (this.document || this.ownerDocument).querySelectorAll(s);
        var i = matches.length;
        while (--i >= 0 && matches.item(i) !== this) {}
        return i > -1;
    };
}

var func = _react.PropTypes.func,
    string = _react.PropTypes.string,
    any = _react.PropTypes.any;

var EventListener = function (_Component) {
    (0, _inherits3.default)(EventListener, _Component);

    function EventListener(props) {
        (0, _classCallCheck3.default)(this, EventListener);

        var _this = (0, _possibleConstructorReturn3.default)(this, (EventListener.__proto__ || (0, _getPrototypeOf2.default)(EventListener)).call(this, props));

        _this.listener = function (event) {
            var _this$props = _this.props,
                target = _this$props.target,
                on = _this$props.on,
                noPrevent = _this$props.noPrevent;

            if (!target || event.target.matches(target)) {
                if (!noPrevent) {
                    event.preventDefault();
                }
                on(event);
            }
        };

        var capture = props.capture,
            once = props.once,
            passive = props.passive;

        _this.state = {
            capture: !!capture,
            once: !!once,
            passive: !!passive
        };
        return _this;
    }

    (0, _createClass3.default)(EventListener, [{
        key: 'componentWillMount',
        value: function componentWillMount() {
            var event = this.props.event;

            document.addEventListener(event, this.listener, this.state);
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            var event = this.props.event;

            document.removeEventListener(event, this.listener, this.state);
        }
    }, {
        key: 'render',
        value: function render() {
            var _props = this.props,
                event = _props.event,
                target = _props.target;

            return _react2.default.createElement(_Empty2.default, { comment: 'On ' + target + ' ' + event + ' listener' });
        }
    }]);
    return EventListener;
}(_react.Component);

EventListener.propTypes = {
    event: string.isRequired,
    target: string,
    on: func.isRequired,
    noPrevent: any,
    capture: any,
    once: any,
    passive: any
};
exports.default = EventListener;
