// react-snippets/Switch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Empty = require('./Empty');

var _Empty2 = _interopRequireDefault(_Empty);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var any = _react.PropTypes.any,
    func = _react.PropTypes.func,
    string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    objectOf = _react.PropTypes.objectOf;

var defaultComment = 'default render condition';

var Switch = function (_Component) {
    (0, _inherits3.default)(Switch, _Component);

    function Switch() {
        (0, _classCallCheck3.default)(this, Switch);
        return (0, _possibleConstructorReturn3.default)(this, (Switch.__proto__ || (0, _getPrototypeOf2.default)(Switch)).apply(this, arguments));
    }

    (0, _createClass3.default)(Switch, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                value = _props.value,
                cases = _props.cases,
                def = _props.def,
                props = _props.props,
                comment = _props.comment;

            var Component = value in cases ? cases[value] : def;

            if ((0, _utils.isComponent)(Component)) {
                return _react2.default.createElement(Component, props);
            }
            if (typeof Component === 'function') {
                return Component(props);
            }
            return _react2.default.createElement(_Empty2.default, { comment: comment });
        }
    }]);
    return Switch;
}(_react.Component);

Switch.propTypes = {
    value: any.isRequired,
    cases: objectOf(func).isRequired,
    def: func,
    props: object,
    comment: string
};
Switch.defaultProps = {
    props: {},
    comment: defaultComment
};
exports.default = Switch;
